File: mdimporter/GetMetadataForFile.c
#include <CoreFoundation/CoreFoundation.h>
#include <CoreServices/CoreServices.h> 
#include <ApplicationServices/ApplicationServices.h>
#include <machine/endian.h>

#include "SOIL/image_DXT.h"	// For DDS_header and flags


// https://developer.apple.com/library/mac/documentation/Carbon/Conceptual/MDImporters/Concepts/WritingAnImp.html
// http://msdn.microsoft.com/en-us/library/bb943991

Boolean GetMetadataForURL(void* thisInterface,
			   CFMutableDictionaryRef attributes, 
			   CFStringRef contentTypeUTI,
			   CFURLRef url)
{

	CGDataProviderRef dataProvider = CGDataProviderCreateWithURL(url);
	if (!dataProvider) return FALSE;
	CFDataRef data = CGDataProviderCopyData(dataProvider);
	CGDataProviderRelease(dataProvider);
	if (!data) return FALSE;

	DDS_header *header = (DDS_header *) CFDataGetBytePtr(data);
	if (CFDataGetLength(data) < sizeof(DDS_header) ||
		memcmp(&header->dwMagic, "DDS ", 4))
	{
		CFRelease(data);
		return FALSE;	/* Not a DDS file! */
	}

	int ncodecs = 0;
	CFTypeRef codecs[3];
	CFStringRef format = NULL;
	if (header->sPixelFormat.dwFlags & DDPF_FOURCC)
		codecs[ncodecs++] = format = CFStringCreateWithBytes(kCFAllocatorDefault, (const UInt8 *) &header->sPixelFormat.dwFourCC, 4, kCFStringEncodingASCII, false);
	else if (header->sPixelFormat.dwFlags & DDPF_RGB)
		codecs[ncodecs++] = (header->sPixelFormat.dwFlags & DDPF_ALPHAPIXELS) ? CFSTR("RGBA") : CFSTR("RGB");
	if (header->sCaps.dwCaps2 & DDSCAPS2_CUBEMAP)
		codecs[ncodecs++] = CFSTR("cubemap");
	if ((header->dwFlags & DDSD_MIPMAPCOUNT) && header->dwMipMapCount)
		codecs[ncodecs++] = CFSTR("mipmaps");
	if (ncodecs)
	{
		CFArrayRef cfcodecs = CFArrayCreate(kCFAllocatorDefault, (const void **) codecs, ncodecs, &kCFTypeArrayCallBacks);
		CFDictionaryAddValue(attributes, kMDItemCodecs, cfcodecs);
		CFRelease(cfcodecs);
		if (format)
			CFRelease(format);
	}

	int height = OSReadLittleInt32(header, offsetof(DDS_header, dwHeight));
	CFNumberRef cfheight = CFNumberCreate(kCFAllocatorDefault, kCFNumberIntType, &height);
	CFDictionaryAddValue(attributes, kMDItemPixelHeight, cfheight);
	CFRelease(cfheight);
	
	int width = OSReadLittleInt32(header, offsetof(DDS_header, dwWidth));
	CFNumberRef cfwidth = CFNumberCreate(kCFAllocatorDefault, kCFNumberIntType, &width);
	CFDictionaryAddValue(attributes, kMDItemPixelWidth, cfwidth);
	CFRelease(cfwidth);

	CFRelease(data);
	return TRUE;
}

File: mdimporter/GetMetadataForFile.m
#import <Cocoa/Cocoa.h>

#import <CoreFoundation/CoreFoundation.h>
#import <CoreServices/CoreServices.h>
#import <ApplicationServices/ApplicationServices.h>

#import "DDS/dds.h"

// https://developer.apple.com/library/mac/documentation/Carbon/Conceptual/MDImporters/Concepts/WritingAnImp.html
// http://msdn.microsoft.com/en-us/library/bb943991


Boolean GetMetadataForURL(void* thisInterface,
			   CFMutableDictionaryRef attributes, 
			   CFStringRef contentTypeUTI,
			   CFURLRef url)
{
    @autoreleasepool
    {

        DDS *dds = [DDS ddsWithURL: (__bridge NSURL*) url];
        if (!dds)
            return FALSE;	/* Not a DDS file! */

        NSMutableDictionary *attrs = (__bridge NSMutableDictionary *)attributes;   // Prefer to use Objective-C

        [attrs setValue:[NSArray arrayWithObject:dds.codec] forKey:(NSString *)kMDItemCodecs];

        int nlayers = 0;
        NSObject *layers[2];
        int ddsCaps2 = dds.ddsCaps2;
        if (ddsCaps2 & DDSCAPS2_VOLUME)
            layers[nlayers++] = @"volume";
        else if ((ddsCaps2 & DDSCAPS2_CUBEMAP_ALLFACES) == DDSCAPS2_CUBEMAP_ALLFACES)
            layers[nlayers++] = @"cubemap";
        else if (ddsCaps2 & DDSCAPS2_CUBEMAP)
            layers[nlayers++] = @"partial cubemap";
        if (dds.mipmapCount > 1)
            layers[nlayers++] = @"mipmaps";
        if (nlayers)
            [attrs setValue:[NSArray arrayWithObjects:layers count:nlayers] forKey:(NSString *)kMDItemLayerNames];

        [attrs setValue:[NSNumber numberWithInt:dds.mainSurfaceWidth]  forKey:(NSString *)kMDItemPixelWidth];
        [attrs setValue:[NSNumber numberWithInt:dds.mainSurfaceHeight] forKey:(NSString *)kMDItemPixelHeight];
        if (dds.bpp)
            [attrs setValue:[NSNumber numberWithInt:dds.bpp] forKey:(NSString *)kMDItemBitsPerSample];

    }
    return TRUE;
}
